core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1523,22 +1523,28 @@ abstract class RDD[T: ClassTag](
    * @return an array of top elements
    */
   def takeOrdered(num: Int)(implicit ord: Ordering[T]): Array[T] = withScope {
-    if (num == 0) {
+    if (num == 0 || this.getNumPartitions == 0) {
       Array.empty
     } else {
-      val mapRDDs = mapPartitions { items =>
-        // Priority keeps the largest elements, so let's reverse the ordering.
-        val queue = new BoundedPriorityQueue[T](num)(ord.reverse)
-        queue ++= collectionUtils.takeOrdered(items, num)(ord)
-        Iterator.single(queue)
-      }
-      if (mapRDDs.partitions.length == 0) {
-        Array.empty
-      } else {
-        mapRDDs.reduce { (queue1, queue2) =>
-          queue1 ++= queue2
-          queue1
-        }.toArray.sorted(ord)
+      this.mapPartitionsWithIndex { case (pid, iter) =>
+        if (iter.nonEmpty) {
+          // Priority keeps the largest elements, so let's reverse the ordering.
+          Iterator.single(collectionUtils.takeOrdered(iter, num)(ord).toArray)
```

Comment:
Will the same benefits when using Scala 2.13?

Reply:
do you mean `collectionUtils.takeOrdered`?
both existing `takeOrdered` and  newly added `mergeOrdered` call the Guava implementations, I think they should be irrelative to the Scala version.

Reply:
I am concerned about the difference between the 'toArray' methods, but this can be ignored. From the source code, `IterableOnceOps.toArray` seems more efficient than `TraversableOnce.toArray`